integration-tests/src/tests/client/features/nonrefundable_transfer.rs

```diff
@@ -0,0 +1,346 @@
+    }
+
+    if deploy_contract {
+        let contract = near_test_contracts::sized_contract(1500 as usize);
+        actions.push(Action::DeployContract(DeployContractAction { code: contract.to_vec() }))
+    }
+
+    let tx_result = execute_transaction_from_actions(env, actions, &signer, receiver.clone());
+
+    let outcome = match &tx_result {
+        Ok(outcome) => outcome,
+        _ => {
+            return tx_result;
+        }
+    };
+
+    if !matches!(outcome.status, near_primitives::views::FinalExecutionStatus::SuccessValue(_)) {
+        return tx_result;
+    }
+
+    let gas_cost = outcome.gas_cost();
+    assert_eq!(sender_pre_balance - deposit - gas_cost, env.query_balance(sender()));
+
+    let receiver_after = env.query_account(receiver);
+    let (receiver_expected_amount_after, receiver_expected_non_refundable_after) = if nonrefundable
+    {
+        (receiver_before_amount, receiver_before_nonrefundable + deposit)
+    } else {
+        (receiver_before_amount + deposit, receiver_before_nonrefundable)
+    };
+
+    assert_eq!(receiver_after.amount, receiver_expected_amount_after);
+    assert_eq!(receiver_after.nonrefundable, receiver_expected_non_refundable_after);
+
+    tx_result
+}
+
+fn delete_account(
+    env: &mut TestEnv,
+    signer: &InMemorySigner,
+) -> Result<FinalExecutionOutcomeView, InvalidTxError> {
+    let actions = vec![Action::DeleteAccount(DeleteAccountAction { beneficiary_id: receiver() })];
+    execute_transaction_from_actions(env, actions, &signer, signer.account_id.clone())
+}
+
+/// Can delete account with non-refundable storage.
+#[test]
+fn deleting_account_with_non_refundable_storage() {
+    let mut env = setup_env();
+    let new_account_id: AccountId = "subaccount.test0".parse().unwrap();
+    let new_account = InMemorySigner::from_seed(
+        new_account_id.clone(),
+        KeyType::ED25519,
+        new_account_id.as_str(),
+    );
+    // Create account with non-refundable storage.
+    // Deploy a contract that does not fit within Zero-balance account limit.
+    let create_account_tx_result = exec_transfer(
+        &mut env,
+        signer(),
+        new_account_id.clone(),
+        NEAR_BASE,
+        true,
+        true,
+        false,
+        true,
+    );
+    create_account_tx_result.unwrap().assert_success();
+
+    // Send some NEAR (refundable) so that the new account is able to pay the gas for its deletion in the next transaction.
+    let send_money_tx_result = exec_transfer(
+        &mut env,
+        signer(),
+        new_account_id.clone(),
+        10u128.pow(20),
+        false,
+        false,
+        false,
+        false,
+    );
+    send_money_tx_result.unwrap().assert_success();
+
+    // Delete the new account (that has 1 NEAR of non-refundable balance).
+    let delete_account_tx_result = delete_account(&mut env, &new_account);
+    delete_account_tx_result.unwrap().assert_success();
+    assert!(!account_exists(&mut env, new_account_id));
```

Comment:
I think there should also be an assertion to check that the beneficiary account received the remaining balance from the deleted account, but none of the non-refundable balance.